qualang_tools/utils/qua_utils.py

```diff
@@ -0,0 +1,78 @@
+import numpy as np
+
+
+def interrupt_on_close(figure, current_job):
+    """
+    Allows to interrupt the execution and free the console when closing the live-plotting figure.
+
+    :param figure: the python figure instance correponding to the live-plotting figure.
+    :param current_job: a ``QmJob`` object (see QM Job API) corresponding to the running job.
+    :return: None.
+    """
+
+    def on_close(event):
+        print("Execution stopped by user!")
+        current_job.halt()
+        event.canvas.stop_event_loop()
+
+    figure.canvas.mpl_connect("close_event", on_close)
+
+
+class result_tool:
+    def __init__(self, job, data_list, mode="wait_for_all"):
+        """
+        API to easily fetch data from the stream processing.
+        **Example**: my_results = result_tool(job=my_job, data_list=["I", "Q"], mode="live")
+
+        :param job: a ``QmJob`` object (see QM Job API) corresponding to the current execution.
+        :param data_list: a list of the result names saved in the stream processing.
+        :param mode: current acquisition mode among ['live', 'wait_for_all'], default is 'wait_for_all'. 'live' will fetch data one by one for all named results for live plotting purposes. 'wait_for_all' will wait until all values were processed for all named results.
+        """
+        if not data_list:
+            raise KeyError("The provided data list is empty.")
+        if mode not in ["live", "wait_for_all"]:
+            raise KeyError(
+                f"Mode '{mode}' is not supported. Supported modes are ['live', 'wait_for_all']"
+            )
+        self.data_list = data_list
+        self.mode = mode
+        self.results = []
+        self.data_handles = []
+        self.res_handles = job.result_handles
+        if mode == "live":
+            for data in self.data_list:
+                if hasattr(self.res_handles, data):
+                    self.data_handles.append(self.res_handles.get(data))
+                    self.res_handles.get(data).wait_for_values(1)

```

Comment:
you just got it a a line above :)

Reply:
I am sure you're right, but I don't get it :)

Reply:
It's not that critical...
You can change the 2nd line to be 
`self.data_handles[-1].wait_for_values(1)`
This would reduce the overhead of doing `self.res_handles.get(data)` twice

Reply:
or 
```
data_handle = self.res_handles.get(data)
self.data_handles.append(data_handle)
data_handle.wait_for_values(1)
```

Reply:
I see, good point indeed !